File: src/main/java/com/teamwizardry/librarianlib/client/SpritesMetadataSection.java
package com.teamwizardry.librarianlib.client;

import java.util.List;

import net.minecraft.client.resources.data.IMetadataSection;

public class SpritesMetadataSection implements IMetadataSection {

	public List<SpriteDefinition> definitions;
	public int width, height;
	
	public SpritesMetadataSection(int width, int height, List<SpriteDefinition> definitions) {
		super();
		this.width = width;
		this.height = height;
		this.definitions = definitions;
	}

	public static class SpriteDefinition {
		public String name;
		public int u, v, w, h;
		public int[] frames;
		
		public SpriteDefinition(String name, int u, int v, int w, int h, int[] animationTimes) {
			super();
			this.name = name;
			this.u = u;
			this.v = v;
			this.w = w;
			this.h = h;
			this.frames = animationTimes;
		}
	}
	
}

File: src/main/java/com/teamwizardry/librarianlib/features/sprite/SpritesMetadataSection.kt
package com.teamwizardry.librarianlib.features.sprite

import net.minecraft.client.resources.data.IMetadataSection
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly
import java.util.Arrays

@SideOnly(Side.CLIENT)
class SpritesMetadataSection(var width: Int, var height: Int, var sprites: List<SpriteDefinition>, var colors: List<ColorDefinition>) : IMetadataSection {
    companion object {
        var registered = false
    }
}

class SpriteDefinition(
    var name: String,
    var u: Int, var v: Int, var w: Int, var h: Int,
    var frames: IntArray, var offsetU: Int, var offsetV: Int,
    var minUCap: Int, var minVCap: Int, var maxUCap: Int, var maxVCap: Int,
    var hardScaleU: Boolean, var hardScaleV: Boolean) {
    constructor(name: String) : this(
        name,
        0, 0, 0, 0,
        intArrayOf(0), 0, 0,
        0, 0, 0, 0,
        false, false
    )

    override fun toString(): String {
        return "SpriteDefinition(" +
            "name='$name', uv=($u,$v), wh=($w,$h), frames=${Arrays.toString(frames)}, offsetUV=($offsetU,$offsetV), " +
            "cap=($minUCap,$minVCap,$maxUCap,$maxVCap), hardScaleUV=($hardScaleU,$hardScaleV)" +
            ")"
    }
}

class ColorDefinition(var name: String, var u: Int, var v: Int) {
    constructor(name: String) : this(name, -1, -1)
}
